engine/src/lib.rs

```diff
@@ -342,6 +342,17 @@ mod contract {
         crate::xcc::set_code_version_of_address(&mut io, &args.address, args.version);
     }
 
+    /// Sets the address for the wNEAR ERC-20 contract. This contract will be used by the
+    /// cross-contract calls feature to have users pay for their NEAR transactions.
+    #[no_mangle]
+    pub extern "C" fn factory_set_wnear_address() {
```

Comment:
I take it that we would have access to this in order to get the wNEAR back as compensation?

Reply:
Not exactly. The reason the engine needs to know this address is to transfer the wNEAR the user approved. The flow is as follows: (1) user approves wNEAR ERC-20 to be spent by the precompile, (2) the precompile transfers the required amount of wNEAR ERC-20 to the engine's implicit address, (3) when creating the xcc promises, the engine includes a `call` to itself in which it uses the exit precompile to send that wNEAR to the user's router account, (4) additionally, a promise is included where the router unwraps its wNEAR into actual NEAR to cover its costs.

Note that the engine never had to use any of its own funds at any point in this flow; it is only facilitating the user in spending their own wNEAR. The exception is when the router account is created for the first time. In this case the engine must loan some NEAR to create the router account before it can successfully use the exit precompile to send the user's own wNEAR. In this case the router refunds the engine the amount it loaned.

Also note that if step (2) fails then the user did not approve enough wNEAR to cover their costs and the flow will terminate (revert) there with an error.

Reply:
Makes sense.